File: actin-clinical/src/main/java/com/hartwig/actin/clinical/feed/patient/PatientEntry.java
package com.hartwig.actin.clinical.feed.patient;

import java.time.LocalDate;

import com.hartwig.actin.clinical.datamodel.Sex;
import com.hartwig.actin.clinical.feed.FeedEntry;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class PatientEntry implements FeedEntry {

    @NotNull
    public abstract String id();

    @NotNull
    @Override
    public abstract String subject();

    public abstract int birthYear();

    @NotNull
    public abstract Sex sex();

    @NotNull
    public abstract LocalDate periodStart();

    @Nullable
    public abstract LocalDate periodEnd();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/patient/PatientEntry.kt
package com.hartwig.actin.clinical.feed.emc.patient

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hartwig.actin.clinical.feed.JacksonSerializable
import com.hartwig.actin.clinical.feed.emc.FeedEntry
import com.hartwig.actin.clinical.feed.emc.FeedSubjectDeserializer
import com.hartwig.actin.datamodel.clinical.Gender
import java.time.LocalDate

@JacksonSerializable
data class PatientEntry(
    @JsonProperty("subject")
    @JsonDeserialize(using = FeedSubjectDeserializer::class)
    override val subject: String,
    @JsonProperty("birth_year")
    val birthYear: Int,
    @JsonProperty("gender")
    val gender: Gender,
    @JsonProperty("period_start")
    val periodStart: LocalDate,
    @JsonProperty("period_end")
    val periodEnd: LocalDate?
) : FeedEntry